From 5b8638f924f91bc581c2ce2d45b8f7f235906e81 Mon Sep 17 00:00:00 2001
From: Saleem Rashid <dev@saleemrashid.com>
Date: Sun, 19 Apr 2020 16:04:58 +0100
Subject: [PATCH 121/134] Fix RAPL sysfs detection

Clear the temporary string stream before testing each sysfs entry, to
test the correct entry names.

Ensure that pl2_index is set, rather than using uninitialized memory
when the short_term window isn't found.

(cherry picked from commit 718d7a676de361ba5765c2f9e454322834a80f99)
Signed-off-by: Paolo Pisati <paolo.pisati@canonical.com>
(cherry picked from commit b8d3dbe0a52ed5e4158de5f88d34ed7ec0efe0a4)
Signed-off-by: Paolo Pisati <paolo.pisati@canonical.com>
---
 src/thd_cdev_rapl.cpp | 11 ++++++++---
 src/thd_cdev_rapl.h   |  5 +++--
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/src/thd_cdev_rapl.cpp b/src/thd_cdev_rapl.cpp
index 90c396a..1f5ed6c 100644
--- a/src/thd_cdev_rapl.cpp
+++ b/src/thd_cdev_rapl.cpp
@@ -126,19 +126,20 @@ int cthd_sysfs_cdev_rapl::get_max_state() {
 int cthd_sysfs_cdev_rapl::rapl_sysfs_valid()
 {
 	std::stringstream temp_str;
-	int found = 0;
+	int found_long_term = 0;
 	int i;
 
 	// The primary control is powercap rapl long_term control
 	// If absent we can't use rapl cooling device
 	for (i = 0; i < rapl_no_time_windows; ++i) {
+		temp_str.str(std::string());
 		temp_str << "constraint_" << i << "_name";
 		if (cdev_sysfs.exists(temp_str.str())) {
 			std::string type_str;
 			cdev_sysfs.read(temp_str.str(), type_str);
 			if (type_str == "long_term") {
 				constraint_index = i;
-				found = 1;
+				found_long_term = 1;
 			}
 			if (type_str == "short_term") {
 				pl2_index = i;
@@ -146,7 +147,7 @@ int cthd_sysfs_cdev_rapl::rapl_sysfs_valid()
 		}
 	}
 
-	if (!found) {
+	if (!found_long_term) {
 		thd_log_info("powercap RAPL no long term time window\n");
 		return THD_ERROR;
 	}
@@ -218,6 +219,10 @@ int cthd_sysfs_cdev_rapl::rapl_update_pl2(int pl2)
 	std::stringstream temp_power_str;
 	int ret;
 
+	if (pl2_index == -1) {
+		thd_log_warn("Asked to set PL2 but couldn't find a PL2 device\n");
+		return THD_ERROR;
+	}
 	temp_power_str << "constraint_" << pl2_index << "_power_limit_uw";
 	ret = cdev_sysfs.write(temp_power_str.str(), pl2);
 	if (ret <= 0) {
diff --git a/src/thd_cdev_rapl.h b/src/thd_cdev_rapl.h
index 79af385..358f3b1 100644
--- a/src/thd_cdev_rapl.h
+++ b/src/thd_cdev_rapl.h
@@ -72,7 +72,7 @@ public:
 			cthd_cdev(_index,
 					"/sys/devices/virtual/powercap/intel-rapl/intel-rapl:0/"), phy_max(
 					0), package_id(package), constraint_index(
-					0), dynamic_phy_max_enable(
+					0), pl2_index(-1), dynamic_phy_max_enable(
 					false), pl0_max_pwr(0), pl0_min_pwr(0), pl0_min_window(
 					0), pl0_max_window(0), pl0_step_pwr(
 					0), bios_locked(false), constrained(
@@ -84,7 +84,8 @@ public:
 			std::string contol_path) :
 			cthd_cdev(_index, contol_path), phy_max(0), package_id(
 					package), constraint_index(
-					0), dynamic_phy_max_enable(false), pl0_max_pwr(
+					0), pl2_index(
+					-1), dynamic_phy_max_enable(false), pl0_max_pwr(
 					0), pl0_min_pwr(
 					0), pl0_min_window(0), pl0_step_pwr(0), bios_locked(
 					false), constrained(
-- 
2.27.0

